sbcl: Small fixes around the phase definitions

equill opened the pull request:
- Remove some debugging output at the start of the `patchPhase` section,
  which appears to have outlived its value.
- Rename `patchPhase` to `postPatch`, to avoid preventing people adding
  patches via the `patches` variable.
- Add `preBuild` and `postBuild` run-hooks to the `buildPhase` section.
- Add `preInstall` and `postInstall` run-hooks to the `installPhase`
  section.

###### Motivation for this change

This replaces PR https://github.com/NixOS/nixpkgs/pull/113132 which was obsoleted by a separate bit of refactoring.

Small improvements to the common functionality for building the SBCL compiler: remove debug output which is now effectively cruft, and make it easier for people to use standard mechanisms to modify this package.


###### Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- [ * ] Tested using sandboxing ([nix.useSandbox](https://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](https://nixos.org/nix/manual/#sec-conf-file) on non-NixOS linux)
- Built on platform(s)
   - [ * ] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
   - N/A - there aren't any tests
- [ * ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
   - "No diff detected"
- [ * ] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [ ] Ensured that relevant documentation is up to date
    - N/A - no documentation to change
- [ * ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

equill committed:
sbcl: Small fixes around the phase definitions

- Remove some debugging output at the start of the `patchPhase` section,
  which appears to have outlived its value.
- Rename `patchPhase` to `postPatch`, to avoid preventing people adding
  patches via the `patches` variable.
- Add `preBuild` and `postBuild` run-hooks to the `buildPhase` section.
- Add `preInstall` and `postInstall` run-hooks to the `installPhase`
  section.

SuperSandro2000 commented:
This is a semi-automatic executed nixpkgs-review with [nixpkgs-review-checks extension](https://github.com/SuperSandro2000/nixpkgs-review-checks). It is checked by a human on a best effort basis and does not build all packages (e.g. lumo, tensorflow or pytorch).
If you have any questions or problems please reach out to SuperSandro2000 on IRC.

Result of `nixpkgs-review pr 113390` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
<summary>
1 package built:
</summary>
<ul>
<li>sbcl</li>
</ul>
</details>

The following issues got detected with the above build packages.
Please fix at least the ones listed with your changed packages:

sbcl:
<details>

Please consider this feature to be alpha.

A substituteInPlace with an unmatched pattern got detected:
```
substituteStream(): WARNING: pattern 'gcc' doesn't match anything in file 'src/runtime/Config.x86-64-darwin'
substituteStream(): WARNING: pattern 'mmacosx-version-min=10.4' doesn't match anything in file 'src/runtime/Config.x86-64-darwin'
```

Please check the offending substituteInPlace for typos or changes in source.
</details>

SuperSandro2000 commented:
If you fix the warning we cleaned up a lot.

r-rmcgibbo commented:
Result of `nixpkgs-review pr 113390` at 89d45049 run on aarch64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>4 packages marked as broken and skipped:</summary>
  <ul>
    <li>acl2</li>
    <li>clfswm</li>
    <li>stumpwm</li>
    <li>stumpwm-git</li>
  </ul>
</details>
<details>
  <summary>1 package failed to build:</summary>
  <ul>
    <li><a href="https://gist.github.com/8fff7dcc571c2a73e8a8d64c115efb51">fricas</a></li></li>
  </ul>
</details>
<details>
  <summary>9 packages built:</summary>
  <ul>
    <li><a href="https://gist.github.com/7f57b8f0d4f8a26b59d2aa8ee12cb257">acl2-minimal</a></li></li>
    <li><a href="https://gist.github.com/29f784748414a684c8d973904dc1aaad">clpm</a></li></li>
    <li><a href="https://gist.github.com/4b4f5d48957bd712fc7ac31753f3c379">maxima</a></li></li>
    <li><a href="https://gist.github.com/c70dc8cbada55f57127360be78ee35d0">pgloader</a></li></li>
    <li><a href="https://gist.github.com/43d4d0eede4ccacbf15ec1fb87ab3896">sbcl</a></li></li>
    <li><a href="https://gist.github.com/230dadf405e216ee1d311356453d78e4">sbcl_2_0_9</a></li></li>
    <li><a href="https://gist.github.com/e2eaa2c472e21d0c987a6ef9718e6fc8">sbcl_2_1_1</a></li></li>
    <li><a href="https://gist.github.com/397ac1b22c2949abc68da8b14c9a08c4">shen-sbcl</a></li></li>
    <li><a href="https://gist.github.com/a3022524d30360eae1836b94ed816cbb">wxmaxima</a></li></li>
  </ul>
</details>

sternenseemann commented on pkgs/development/compilers/sbcl/common.nix:
```diff
@@ -22,11 +22,7 @@ stdenv.mkDerivation rec {
   buildInputs = [texinfo];
 
-  patchPhase = ''
-    echo '"${version}.nixos"' > version.lisp-expr
```
Maybe the `.nixos` suffix for the version was intentional? The [sbcl build system uses this file](https://github.com/sbcl/sbcl/blob/545d04f6ac15cdf678f97015c11efd9f92d8afee/generate-version.sh#L14-L26).

equill replied:
I see what you mean.
I was about to revert that part, but it didn't seem to cause a problem with the build, so I checked how it's used.
In (doc/GIT-WORKFLOW.MD)[https://github.com/sbcl/sbcl/blob/master/doc/GIT-WORKFLOW.md] it says that this ceased to be useful after switching from CVS to a decentralised, branching VCS.
Thus, "in Git workflow, version.lisp-expr no longer exists in the repository, nor is the version number prepended to commit messages." 
There's more explanation in that file, but the upshot is that its use is deprecated if you're building from `git`.
`doc/GIT-WORKFLOW.md` is about 10 years old, and `sbc/default.nix` dates back 11 years to 2009, so I think this is a historical artifact that's survived until now simply because it didn't break anything.

sternenseemann replied:
This is what it is set to in the relase tarball:

```
tar -xjOf /nix/store/k6gxslva6zvdgl0wbmgpgnh4qv55h3p3-sbcl-2.0.8-source.tar.bz2 sbcl-2.0.8/version.lisp-expr
;;; This file is auto-generated using generate-version.sh. Every time
;;; you re-run make.sh, this file will be overwritten if you are
;;; working from a Git checkout.
"2.0.8"
```

So it still holds that we overwrite this to say `2.0.8.nixos` which means we change something about the build process, rather than just removing unused code.

equill replied:
It's not a hill I'm eager to die on, so I've reinstated that line.

sternenseemann replied:
I guess an opinion from @7c6f434c would be interesting, as they are mantaining `sbcl`.

7c6f434c replied:
I think we change some low-level system stuff (like uname call) so it makes sense to have a suffix, and we do not pass git to SBCL build, so it is not clear we should drop the version (and start-up banner does end up displaying this version, so it is not completely moot)

equill commented:
I'm confused by the issue with substituting 'mmacosx-version-min=10.4' - that string doesn't exist in that file, so the substitution failure is valid. It hasn't existed in that file at least since 2.0.8, though it's present in the x86 (non-64) and ppc variants of this file.
What confuses me is why it's turned up as a warning now, when it didn't (and I assume should have) previously. I'm guessing it's because the Nixos build system only just acquired this feature.
I'll remove that substitution as another historical artifact, unless anybody points out what I missed.

Now double-checking the gcc/cc issue, which I suspect to be the same thing, and then I'll see if I can figure out what broke `fricas`.

equill committed:
sbcl: Remove substution of 'gcc' with 'cc'

This substitution produces a nix-build warning, because the target
string was removed in commit 7b225df9a492977b8fff0948ce93c5ab31766f64
in 2016.

Remove the substition from the `postPatch` phase in `sbcl/common.nix`.

equill committed:
sbcl: Remove substitution of mmacosx-verson-min

This substition produces a nixos-build error, because the target string
was removed in commit 323891ea5964d54aba2d272ddddbf748ce353176 in 2018.

Remove the substition from the `postPatch` phase of `sbcl/common.nix`.

equill commented:
A bit more forensic archaeology confirmed that both those substitutions are historical cruft, and can safely be removed. So I removed them.
If `fricas` is still broken, I'll take a closer look at that.

equill commented:
Also, sorry about annoying you by referring to that as a 'refactor'. I admit that I didn't look closely at the amount of change involved.

equill committed:
sbcl: Resume creating version.lisp-expr file

Reverse the recent commit that removed this line.
